python/lib/core/dmod/core/dataset.py

```diff
@@ -177,19 +177,42 @@ def __init__(self, manager: DatasetManager = None, **kwargs):
                 raise ValueError(f"Expected UUID got {type(manager.uuid)}")
             self.manager_uuid = manager.uuid
 
+    def __eq__(self, other):
+        return (
+            isinstance(other, Dataset)
+            and self.access_location == other.access_location
+            and self.category == other.category
+            and self.created_on == other.created_on
+            and self.data_domain == other.data_domain
+            and self.dataset_type == other.dataset_type
```

Comment:
Thanks for bringing up these questions! I had similar questions myself. I went back in the [git history to the first point before the `pydantic` refactor work](https://github.com/NOAA-OWP/DMOD/blob/ac8b0474497b479cdedcb394b8263a1f21f42acb/python/lib/core/dmod/core/dataset.py#L108-L112) and copied and pasted the `__eq__` implementation from that point. I thought that would at least be a good starting point.

I am not sure if we should compare `uuid`'s or not? I lean toward yes, but I could hear the argument for no. Our idea of a dataset has slightly changed since pre-`pydantic` refactor as we now use, more or less, aggregate datasets for everything but forcing data for a simulation. I think that change in thinking was probably why they were not compared in the previous version of the code when the thinking was to determine suitable existing datasets that could be used to satisfy a given simulation request? Is that fair and what are your thoughts?

Reply:
> Our idea of a dataset has slightly changed since pre-pydantic refactor as we now use, more or less, aggregate datasets for everything but forcing data for a simulation.

I don't think our abstract idea has change; we just evolved another more practical data format.  

Here's what I suggest:
- `expires`: test both are in the future
  - disagreeing about how long data is good for doesn't imply disagreeing on whether your are talking about the same data, as long as everyone agrees the data is still good
- `derived_from`: optionally test for equality
  - different _explicit_ origins implies different data, but lack of information about origin does not
- `derivations`: ignore
  - doesn't really say anything about "this" data 
- `last_updated`: optionally test for equality
  - if one side says things have updated more recently than the other, chances are these are not truly talking about the same data
- `uuid`: ??? ... for now, optionally test for equality, but this needs it's own issue for follow-up
  - I thought more about this, and I'm less sure of what to do; does it make sense to have a universally unique identifier if it is not universally unique (i.e., if there are ever two non-equal objects with the same value, even if the value is `None`)?

Reply:
Is it okay if `expires` is None on both datasets?

Reply:
Ah, yes, I should have clarified, but I think that's fine also.  `None` implies the dataset is not temporary.   Certainly if both are `None`, this fits with equality.  And I don't think disagreeing on whether the dataset is temporary should necessarily imply looking at different data (again, as long as the expire time hasn't passed yet).

Reply:
Thanks for clarifying!